### PR TITLE
[DRAFT] Remove State#onEnter

### PR DESCRIPTION
**Pull request by zentol**
ping @rmetzger @tillrohrmann 

Based on #14921.

`onEnter` logic was moved into the constructors in 7116435 .
Tests were adjusted in 6bfe3a8 , including the remove of `onEnter` calls and adjusting the order of statements.
Finally, 65ba2f0 disables all tests that are currently failing. Overall the main theme is `State` methods being called without `onEnter` ever having been called. This results in us testing undefined behavior in quite a few tests.
(e.g., you'd call `suspend` on `Canceling` but job wasn't even canceled yet, so you're practically testing some prior state (like `Executing`))

This generally manifests in either 
- an NPE in the MockExecutionGraph because some methods required for the execution have not been setup
- a state transitions unexpectedly being initiated and possible being completed right away

For DeclarativeScheduler#transitionTo I for the time being pass a `Supplier<State>` and the corresponding class separately, instead of wrapped into a factory because it does not really seem worth the hassle. (you still have to pass a supplier to some factory, and we'd one for each state)

### CODE REVIEW

**Review by tillrohrmann — COMMENTED** [review]
I like the proposal @zentol. As you have already found several places where we test impossible situations, I think this is a good improvement as it makes the contract easier for devs. +1 for fully implementing this improvement.

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeScheduler.java [hunk 1]
```diff
@@ -887,19 +897,22 @@ public void runIfState(State expectedState, Runnable action, Duration delay) {
...
 
-    private void transitionToState(State newState) {
-        if (state != newState) {
-            LOG.debug(
-                    "Transition from state {} to {}.",
-                    state.getClass().getSimpleName(),
-                    newState.getClass().getSimpleName());
+    @VisibleForTesting
+    <S extends State> void transitionToState(
+            Supplier<S> stateTransition, Class<S> targetStateClass) {
```
[the remark this answers is not in the document]
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeScheduler.java:902]: What exactly are you thinking of? Just some generic Pojo that you'd create like this:
```
new StateFactory(
                () -> new WaitingForResources(this, LOG, desiredResources, Duration.ofSeconds(10)),
                WaitingForResources.class)
```

or more like an interface with explicit implementations for each state:
```
class CreatedStateFactory implements StateFactory<Created> {

    public Class<Created> getStateClass() { return Created.class; }

    // supplier variant:
    // usage: new StateFactory(() -> new WaitingForResources(this, LOG, desiredResources, Duration.ofSeconds(10))
    public Created getState() { return this.supplier.get(); }

    // constructor variant
    // usage new StateFactory(this, LOG, desiredResources, Duration.ofSeconds(10)
    public Created getState() { return new WaitingForResources(this.scheduler, this.log, this.desiredResources, Duration.ofSeconds(10); }
}
```

Overall, I'm not sold yet on either option; it feels like wed'd be adding code without really changing things all that much on the practical side of things.

**Comment by rmetzger** [comment]
👍 for this change!

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeScheduler.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeScheduler.java:902]: Admittedly it does not change a lot but I would go with the second option using a constructor. That way we wouldn't have to introduce a generic type parameter for `transitionToState`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeScheduler.java [hunk 2]
```diff
@@ -887,19 +897,22 @@ public void runIfState(State expectedState, Runnable action, Duration delay) {
...
     // ----------------------------------------------------------------
 
-    private void transitionToState(State newState) {
-        if (state != newState) {
-            LOG.debug(
-                    "Transition from state {} to {}.",
-                    state.getClass().getSimpleName(),
-                    newState.getClass().getSimpleName());
+    @VisibleForTesting
+    <S extends State> void transitionToState(
```
Comment (rmetzger) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeScheduler.java:901]: why did you change the semantics of "transition to same state" from "ignore" to "fail"?
